gaphor/ui/toolbox.py

```diff
@@ -31,60 +31,140 @@ class Toolbox(Gtk.ToolPalette):
         ),
     ]
 
-    __gsignals__ = {
-        "toggled": (
-            GObject.SignalFlags.RUN_FIRST,
-            None,
-            (GObject.TYPE_STRING, GObject.TYPE_INT),
-        )
-    }
+    title = _("Toolbox")
+    placement = ("left", "diagrams")
 
+    component_registry = inject("component_registry")
+    main_window = inject("main_window")
     properties = inject("properties")
 
-    def __init__(self, toolboxdef):
-        """Create a new Toolbox instance.
-
-        ToolItemGroups and ToolItems are created using the menu_factory and
-        based on the toolboxdef definition.
+    menu_xml = """
+      <ui>
+        <menubar name="mainwindow">
+          <menu action="diagram">
+            <separator/>
+            <menuitem action="reset-tool-after-create" />
+            <separator/>
+          </menu>
+        </menubar>
+      </ui>
+    """
 
-        """
-        GObject.GObject.__init__(self)
+    def __init__(self, toolbox_actions=TOOLBOX_ACTIONS):
+        self._toolbox = None
+        self._toolbox_actions = toolbox_actions
+        self.action_group = build_action_group(self)
+        self.action_group.get_action("reset-tool-after-create").set_active(
+            self.properties.get("reset-tool-after-create", True)
+        )
         self.buttons = []
         self.shortcuts = {}
-        self._construct(toolboxdef)
-
-    def toolbox_button(self, action_name, stock_id):
-        button = Gtk.ToolButton.new_from_stock(stock_id)
-        button.action_name = action_name
-        button.set_use_drag_window(True)
-
-        # Enable Drag and Drop
-        button.drag_source_set(
-            Gdk.ModifierType.BUTTON1_MASK,
-            self.DND_TARGETS,
-            Gdk.DragAction.COPY | Gdk.DragAction.LINK,
-        )
-        button.drag_source_set_icon_stock(stock_id)
-        button.connect("drag-data-get", self._button_drag_data_get)
 
-        return button
-
-    def _construct(self, toolboxdef):
-        shortcuts = self.shortcuts
-        for title, items in toolboxdef:
+    def open(self):
+        widget = self.construct()
+        self.main_window.window.connect_after(
+            "key-press-event", self._on_key_press_event
+        )
+        self.component_registry.register_handler(self._on_diagram_page_change)
+        return widget
+
+    def close(self):
+        if self._toolbox:
+            self._toolbox.destroy()
+            self._toolbox = None
+        self.component_registry.unregister_handler(self._on_diagram_page_change)
+
+    def construct(self):
+        def toolbox_button(action_name, stock_id, label, shortcut):
+            button = Gtk.ToggleToolButton.new()
+            button.set_icon_name(stock_id)
+            button.action_name = action_name
+            if label:
+                button.set_tooltip_text("%s (%s)" % (label, shortcut))
+
+            # Enable Drag and Drop
+            inner_button = button.get_children()[0]
```

Comment:
This looks like the magic that made this work: why / how does the ToggleToolButton have a child button?

Reply:
It took me some digging into the Gtk code 😜